File: python/6.py
import re
import sys
from pathlib import Path
from typing import Any, List, Optional, Tuple

sys.path.append(str(Path(__file__).parent.parent.parent))
from Modules.timer import Timer
from collections import defaultdict


def part1(data):
    """ 2020 Day 6 Part 1

    >>> part1(['abc', '', 'a', 'b', 'c', '', 'ab', 'ac', '', 'a', 'a', 'a', 'a', '', 'b', ''])
    11
    """

    count = 0
    group = defaultdict(lambda: 0)
    inGroup = 0
    for line in data:
        if len(line) == 0:
            count += len(group)

            group = defaultdict(lambda: 0)
            inGroup = 0
            continue
        
        inGroup += 1
        for c in line:
            group[c] += 1

    count += len(group)

    return count


def part2(data):
    """ 2020 Day 6 Part 2

    >>> part2(['abc', '', 'a', 'b', 'c', '', 'ab', 'ac', '', 'a', 'a', 'a', 'a', '', 'b', ''])
    6
    """

    count = 0
    group = defaultdict(lambda: 0)
    inGroup = 0
    for line in data:
        if len(line) == 0:
            count += len([g for g in list(group.keys()) if group[g] == inGroup])

            group = defaultdict(lambda: 0)
            inGroup = 0
            continue
        
        inGroup += 1
        for c in line:
            group[c] += 1
    
    count += len([g for g in list(group.keys()) if group[g] == inGroup])

    return count


def main(input_path: Optional[Path | str]=None, verbose: bool=False) -> Tuple[Tuple[Any, float]]:
    if not input_path:
        if not (input_path := sys.argv[1] if len(sys.argv) > 1 else None):
            year, day = re.findall(r'\d+', str(__file__))[-2:]
            input_path = Path(Path(__file__).parent.parent.parent, "Inputs", f"{year}_{day}.txt")
    
    with open(input_path, encoding='UTF-8') as f:
        data = [line.strip('\n') for line in f.readlines()]

    with Timer() as p1_time:
        p1 = part1(data)

    if verbose:
        print(f"\nPart 1:\nNumber of questions answered yes: {p1}\nRan in {p1_time.elapsed:0.4f} seconds")

    with Timer() as p2_time:
        p2 = part2(data)

    if verbose:
        print(f"\nPart 2:\nNumber of questions answered yes by everyone in a group: {p2}\nRan in {p2_time.elapsed:0.4f} seconds")

    return [(p1, p1_time.elapsed), (p2, p2_time.elapsed)]


if __name__ == "__main__":
    main(verbose=True)

File: c/6.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>
#include <ctype.h>
#define defaultInput "../../Inputs/2020_6.txt"
#define dataLine 30

int findLines(char *fileName) {
	int numLines = 0;
	char textRead[dataLine];

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	// Check if the file exists or not
    if (inFile == NULL) {
        return -1;
    }

	while(fgets(textRead, dataLine, inFile)) {
        ++numLines;
	}

	fclose(inFile);

	return numLines;
}

void groupSize(int *groups, char *fileName) {
    int group = 0, i = 0;
	char textRead[dataLine];

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	while(fgets(textRead, dataLine, inFile)) {
        if (textRead[0] == '\n') {
            groups[i] = group;
            i++;
            group = 0;
        } else {
            group++;
        }
	}
    
	fclose(inFile);
    groups[i] = group;

    return;
}

int charCount(char *str, int ch) {
    int count = 0;
    for (int i = 0; str[i] != '\0'; ++i) {
        if (ch == str[i])
            ++count;
    }

    return count;
}

int part1(char *fileName) {
    int count = 0;
	char textRead[dataLine], ans[dataLine];
	for (int i = 0; i < dataLine; i++) {
        ans[i] = (i == 0) ? '\n' : ' ';
    }

	// Open the file
	FILE *inFile = fopen(fileName, "r");

	while(fgets(textRead, dataLine, inFile)) {
        if (textRead[0] == '\n') {
            int i = 0;
            while (ans[i] != '\n') {
                count++;
                i++;
            }

            for (i = 0; i < dataLine; i++) {
                ans[i] = (i == 0) ? '\n' : ' ';
            }

            continue;
        }

        int i = 0;
        while (textRead[i] != '\n') {
            char *p;
            char temp = textRead[i];
            char temps[dataLine];
            strcpy(temps, ans);

            p = strchr(ans, textRead[i]);

            if (p == NULL) {
                p = strchr(ans, '\n');

                *p = temp;
                *(p + 1) = '\n';
            }

            i++;
        }
	}

	fclose(inFile);

    int i = 0;
    while (ans[i] != '\n') {
        count++;
        i++;
    }
    return count;
}

int part2(int *groups, char *fileName) {
    int count = 0, lineNum = 0;
	char textRead[dataLine], *ans;

	// Open the file
	FILE *inFile = fopen(fileName, "r");

    ans = (char *) malloc((groups[0] * dataLine) + 1);
    ans[0] = '\n';

	while(fgets(textRead, dataLine, inFile)) {
        if (textRead[0] == '\n') {
            for (int i = 97; i<= 122; i++) {
                if (charCount(ans, i) == groups[lineNum]) {
                    count++;
                }
            }

            lineNum++;
            free(ans);
            ans = (char *) malloc((groups[lineNum] * dataLine) + 1);
            ans[0] = '\n';
        } else {
            if (ans[0] == '\n') {
                strcpy(ans, textRead);
            } else {
                strcat(ans, textRead);
            }
        }
	}

	fclose(inFile);
    for (int i = 97; i<= 122; i++) {
        if (charCount(ans, i) == groups[lineNum]) {
            count++;
        }
    }

    return count;
}

int main (int argc, char *argv[]) {
    char *inputPath = defaultInput;
    if (argc > 1) {
        inputPath = argv[1];
    }

	int numLines, count = 0, max;

	numLines = findLines(inputPath);

	if (numLines == -1) {
        printf("Error: Could not read input file. Quitting\n");
        return -1;
	}

    int groups[numLines];
    groupSize(&groups[0], inputPath);

    clock_t t; 
    t = clock(); 
    int p1 = part1(inputPath);
    t = clock() - t; 
    double t_p1 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 1:\nNumber of questions answered yes: %d\nRan in %f seconds\n", p1, t_p1);

    t = clock();
    int p2 = part2(&groups[0], inputPath);
    t = clock() - t;
    double t_p2 = ((double)t) / CLOCKS_PER_SEC;
    printf("\nPart 2:\nNumber of questions answered yes by everyone in a group: %d\nRan in %f seconds\n", p2, t_p2);

	return 0;
}
